### PR TITLE
[Merged by Bors] - Rework s3 configuration

### PR DESCRIPTION
**Pull request by maltesander**
## Description

closes https://github.com/stackabletech/hive-operator/issues/138

integration test: https://github.com/stackabletech/integration-tests/pull/239

Currently requires a certain secret operator pr:
`helm install secret-operator stackable-test/secret-operator --version=0.3.0-pr125`

### TODO
- [x] Use operator-rs tag 0.21.0

## Review Checklist
- [ ] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator-binary/src/controller.rs [hunk 1]
```diff
@@ -354,13 +421,45 @@ fn build_metastore_rolegroup_statefulset(
...
 
     for (property_name_kind, config) in metastore_config {
         match property_name_kind {
             PropertyNameKind::Env => {
+                if let Some(S3ConnectionSpec {
+                    secret_class: Some(secret_name),
+                    ..
+                }) = s3_connection
+                {
+                    container_builder.add_env_var_from_source(
```
Comment (sbernauer) [inline rust/operator-binary/src/controller.rs:437]: What do you think about using `add_env_var_from_secret` from https://github.com/stackabletech/operator-rs/blob/d8687081325924ee68069f39b873f1f25bf0ed09/src/builder/pod/container.rs#L86?

File: rust/operator-binary/src/controller.rs [hunk 2]
```diff
@@ -268,15 +301,49 @@ fn build_metastore_rolegroup_config_map(
...
+                            ssl_enabled = MetaStoreConfig::S3_SSL_ENABLED);
+                        true
+                    } else {
+                        false
+                    };
+
+                    data.insert(
+                        MetaStoreConfig::S3_SSL_ENABLED.to_string(),
+                        Some(use_tls.to_string()),
+                    );
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/controller.rs:334]: yeah i like that thanks.

File: rust/operator-binary/src/controller.rs (hunk 1 above)
Comment (maltesander) [reply rust/operator-binary/src/controller.rs:437]: yep way nice, did not see that!

File: rust/operator-binary/src/controller.rs [hunk 3]
```diff
@@ -130,6 +149,17 @@ pub async fn reconcile_hive(hive: Arc<HiveCluster>, ctx: Context<Ctx>) -> Result
     let client = &ctx.get_ref().client;
     let hive_version = hive_version(&hive)?;
 
+    let s3_connection_def: &Option<S3ConnectionDef> = &hive.spec.s3;
+    let s3_connection_spec: Option<S3ConnectionSpec> = if let Some(s3) = s3_connection_def {
```
Comment (sbernauer) [inline rust/operator-binary/src/controller.rs:153]: Why can't we have stable async closures for .map()-functions? :(

**Review by sbernauer — APPROVED** [review]
LGTM. Some sort of integration-test would be great

**Comment by maltesander** [comment]
> LGTM. Some sort of integration-test would be great

Yeah i wanna try to create something today, or at least integrate that in the trino one.

File: rust/operator-binary/src/controller.rs [hunk 4]
```diff
@@ -468,13 +594,21 @@ fn build_metastore_rolegroup_statefulset(
...
                 .add_volume(Volume {
-                    name: "conf".to_string(),
+                    name: "config".to_string(),
                     config_map: Some(ConfigMapVolumeSource {
                         name: Some(rolegroup_ref.object_name()),
                         ..ConfigMapVolumeSource::default()
                     }),
                     ..Volume::default()
                 })
+                .add_volume(Volume {
```
Comment (sbernauer) [inline rust/operator-binary/src/controller.rs:604]: builders. builds everywhere ^^
https://github.com/stackabletech/operator-rs/blob/ac3f79b354d62a6d769e7f23106763455c322f04/src/builder/pod/volume.rs#L78-L88

**Review by sbernauer — APPROVED** [review]

**Comment by sbernauer** [comment]
When i run the tests with `hive=0.6.0-pr179` i get the following expectation. Can you reproduce that?
```
    logger.go:42: 09:02:43 | smoke/30-prepare-test-metastore | running command: [sh -c kubectl exec -n $NAMESPACE test-metastore-0 -- python /tmp/test_metastore.py -n $NAMESPACE]
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore | Traceback (most recent call last):
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore |   File "/tmp/test_metastore.py", line 71, in <module>
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore |     hive_client.create_external_table(table(database_name, s3_test_table_name, "s3a://test/"))
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore |   File "/root/.local/lib/python3.10/site-packages/hive_metastore_client/hive_metastore_client.py", line 218, in create_external_table
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore |     self.create_table(table)
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore |   File "/root/.local/lib/python3.10/site-packages/thrift_files/libraries/thrift_hive_metastore_client/ThriftHiveMetastore.py", line 2633, in create_table
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore |     self.recv_create_table()
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore |   File "/root/.local/lib/python3.10/site-packages/thrift_files/libraries/thrift_hive_metastore_client/ThriftHiveMetastore.py", line 2659, in recv_create_table
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore |     raise result.o3
    logger.go:42: 09:02:45 | smoke/30-prepare-test-metastore | thrift_files.libraries.thrift_hive_metastore_client.ttypes.MetaException: MetaException(message='Got exception: java.io.InterruptedIOException doesBucketExist on test: com.amazonaws.AmazonClientException: No AWS Credentials provided by BasicAWSCredentialsProvider EnvironmentVariableCredentialsProvider InstanceProfileCredentialsProvider : com.amazonaws.SdkClientException: Failed to connect to service endpoint: ')

**Review by sbernauer — APPROVED** [review]
Implementation LGTM :+1:
We still need to sort out minor stuff such as release operator-rs

**Review by sbernauer — APPROVED** [review]
can now be merged

**Pull request closed without merging**